### PR TITLE
fix:support await fastify.view mode with art-template

### PR DESCRIPTION
**Pull request by momofan**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by Eomm** [comment]
Would you mind to add a unit test for this case?

**Comment by momofan** [comment]
> Would you mind to add a unit test for this case?

OK, I added it.

**Comment by Eomm** [comment]
I'm checking since it is related to #127

File: index.js [hunk 1]
```diff
@@ -49,7 +49,10 @@ function fastifyView (fastify, opts, next) {
 
     const promise = new Promise((resolve, reject) => {
       renderer.apply({
-        getHeader: () => {},
+        getHeader: () => {
```
Comment (Eomm) [inline index.js]: Could you revert this change?
The previous version return `undefined` that is a falsy value so the `if condition` will be false aswell

Comment (momofan) [reply index.js]: OK I got it.

File: index.js [hunk 2]
```diff
@@ -241,7 +241,10 @@ function fastifyView (fastify, opts, next) {
 
     try {
       const html = render(page, data)
-      this.header('Content-Type', 'text/html; charset=' + charset).send(html)
+      if (!this.getHeader('content-type')) {
+        this.header('Content-Type', 'text/html; charset=' + charset)
+      }
```
Comment (mcollina) [inline index.js]: I don’t understand why this change would fix this, would you mind adding a comment or explaining it in the PR?

Comment (momofan) [reply index.js]: this.header('Content-Type', 'text/html; charset=' + charset)

This statement returns undefined
undefined.send(html) 
This execution will report error

Comment (jsumners) [reply index.js]: Are you sure? `.header` should only ever return the `Reply` instance — https://github.com/fastify/fastify/blob/639298b38ae2e16ee3c6e951614145a892964412/lib/reply.js#L176

Comment (momofan) [reply index.js]: image
fastify.view will take this logic

Comment (mcollina) [reply index.js]: A better fix is to have `header` return the instance in that mock.

Comment (Eomm) [reply index.js]: That condition is written also for others template's engine, so it would be better to use always the same logic

Comment (momofan) [reply index.js]: so the final solution is？Hope to solve it as soon as possible，thank you

Comment (momofan) [reply index.js]: so the final solution is？I hope to use this function in a few days, thanks!

Comment (jsumners) [reply index.js]: My opinion is that these implementations are wrong:

https://github.com/fastify/point-of-view/blob/b7345f0601ec0e789cae535cf96bc14ac42ac04d/index.js#L52-L53

They are replicating the core `Reply` methods and should behave accordingly. In fact, I'm not sure why the `view` decoration is on the server object. Views should be rendered in the context of a request, and in particular the context of sending a reply. So I would think the `view` decoration should be `fastify.decorateReply('view', function () {})`. Which would mean the following would be an outline of a solution:

```js
fastify.decorateReply('view', function () {
  // other details omitted
  renderer.apply({getHeader, header}, args)

  function getHeader (key) {
    return this.getHeader(key)
  }

  function header (key, val) {
    return this.header(key, val)
  }
})
```

Comment (florianb) [reply index.js]: I guess i'd try to create a corresponding PR since this issue is blocking for us and i'd like to fix that without monkeypatching. I'm sure it would take me a few days to make it.

Comment (Eomm) [reply index.js]: @mcollina do you agree with my comment?

Comment (florianb) [reply index.js]: In addition to @jsumners suggestion i'd say `fastify`as well as the `reply` object should be decorated. It might be necessary to customize the reply further or use the views to f.e. generate email-content as well.

Without any offences - it seems like the code-structure could be optimized (there are a lot of redundancies) in general to make future maintenance easier and to clean up some individual behaviors of the template engines.

**Comment by Eomm** [comment]
Following the [comment not blocking](https://github.com/orgs/fastify/teams/fastify/discussions/53) I'm merging this and I will open a new issue to clear the code to implement the fluent usage and remove the `if()` statements in all the templates (not only the art one)
